File: models/src/main/java/io/xj/hub/meme/ParseAnti.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.hub.meme;

import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 Meme Matcher for Anti-Memes
 <p>
 Parse any meme to test if it's valid, and extract its features
 <p>
 Artist can add !MEME values into Programs https://www.pivotaltracker.com/story/show/176474073
 */
class ParseAnti {
  static final Pattern rgx = Pattern.compile("^!(.+)$");
  final String body;
  final boolean isValid;

  ParseAnti(String raw) {
    Matcher matcher = rgx.matcher(raw);

    if (!matcher.find()) {
      body = raw;
      isValid = false;
      return;
    }

    body = matcher.group(1);
    if (java.util.Objects.isNull(body) || body.length() == 0) {
      isValid = false;
      return;
    }

    isValid = true;
  }

  public static ParseAnti fromString(String raw) {
    return new ParseAnti(raw);
  }

  public boolean isViolatedBy(ParseAnti target) {
    return (isValid && !target.isValid && Objects.equals(body, target.body)) ||
      (!isValid && target.isValid && Objects.equals(body, target.body));
  }

  public boolean isAllowed(List<ParseAnti> memes) {
    for (var meme : memes)
      if (isViolatedBy(meme)) return false;
    return true;
  }
}

File: engine/src/entities/meme/ParseAnti.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/entities/meme/ParseAnti.h"

using namespace XJ;

ParseAnti::ParseAnti(const std::string &raw) {
  std::smatch matcher;
  isValid = std::regex_search(raw, matcher, rgx);

  if (!isValid) {
    body = raw;
    return;
  }

  body = matcher[1].str();
  if (body.empty()) {
    isValid = false;
    return;
  }
}

ParseAnti ParseAnti::fromString(const std::string &raw) {
  return ParseAnti(raw);
}

bool ParseAnti::isViolatedBy(const ParseAnti &target) {
  return (isValid && !target.isValid && body == target.body) ||
         (!isValid && target.isValid && body == target.body);
}

bool ParseAnti::isAllowed(const std::vector<ParseAnti> &memes) {
  return std::all_of(memes.begin(), memes.end(), [this](const ParseAnti &meme) {
    return !isViolatedBy(meme);
  });
}
